include/stdexec/execution.hpp

```diff
@@ -3236,51 +3285,113 @@ namespace stdexec {
+      }
+
+      /// @post The started bit is set in the shared state's ref count, OR the __waiters_ list
+      /// is set to the known "tombstone" value indicating completion.
+      void __try_start() noexcept {
+        // With the split algorithm, multiple split senders can be started simultaneously, but
+        // only one should start the underlying async operation. Set the low-order "started" bit
+        // in the shared state's ref count. If the previous value was even, then we are the first
+        // consumer to start, so start the underlying async operation.
+        auto __old_ref_count = __ref_count_.fetch_or(__started_bit, std::memory_order_relaxed);
+        if (__started_bit == (__old_ref_count & __started_bit)) {
+          // Already started, do nothing
+          return;
+        }
 
+        // If stop has already been requested when we're starting the underlying the operation,
+        // don't start it; complete with set_stopped immediately.
         if (__stop_source_.stop_requested()) {
-          // 1. resets __head to completion state
-          // 2. notifies waiting threads
-          // 3. propagates "stopped" signal to `out_r'`
-          __notify();
-        } else {
-          stdexec::start(__op_state2_);
+          // 1. Sets __waiters_ to a known "tombstone" value
+          // 2. Notifies all the waiters that the operation has stopped
+          // 3. Clears the "started" bit.
+          __notify_waiters();
+          return;
         }
+
+        stdexec::start(__op_state2_);
       }
 
-      // This is called when the shared async operation completes:
-      void __notify() noexcept {
-        void* const __completion_state = static_cast<void*>(this);
-        void* const __old = __head_.exchange(__completion_state, std::memory_order_acq_rel);
-        auto* __state = static_cast<__local_state_base*>(__old);
+      /// @brief This is called when the shared async operation completes.
+      /// @post __waiters_ is set to a known "tombstone" value.
+      template <class _Tag, class... _As>
+      void __complete(_Tag, _As&&... __as) noexcept {
+        try {
+          using __tuple_t = __decayed_tuple<_Tag, _As...>;
+          __results_.template emplace<__tuple_t>(_Tag(), static_cast<_As&&>(__as)...);
+        } catch (...) {
+          using __tuple_t = __decayed_tuple<set_error_t, std::exception_ptr>;
+          __results_.template emplace<__tuple_t>(set_error, std::current_exception());
+        }
+
+        __notify_waiters();
+      }
+
+      /// @brief This is called when the shared async operation completes.
+      /// @post __waiters_ is set to a known "tombstone" value.
+      void __notify_waiters() noexcept {
+        __waiters_t __waiters_copy{__local_state_base::__get_tombstone()};
+
+        // Set the waiters list to a known "tombstone" value that we can check later.
+        if (std::unique_lock __lock{__mutex_}) { // always true
+          __waiters_.swap(__waiters_copy);
+        }
```

Comment:
Could use `std::lock_guard` here as a simpler facility as you don't need manual unlock capability.
```suggestion
        {
          std::lock_guard __lock{__mutex_};
          __waiters_.swap(__waiters_copy);
        }
```